File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/blecommand/BleCommand.java
package info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.comm.blecommand;

import org.jetbrains.annotations.NotNull;

public abstract class BleCommand {
    private final byte[] data;

    public BleCommand(@NotNull BleCommandType type) {
        this.data = new byte[]{type.getValue()};
    }

    public BleCommand(@NotNull BleCommandType type, @NotNull byte[] payload) {
        int n = payload.length + 1;
        this.data = new byte[n];
        this.data[0] = type.getValue();
        System.arraycopy(payload, 0, data, 1, payload.length);
    }

    public byte[] asByteArray() {
        return this.data;
    }
}

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/command/BleCommand.kt
package info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.comm.command

import info.nightscout.androidaps.utils.extensions.toHex

open class BleCommand(val data: ByteArray) {

    constructor(type: BleCommandType) : this(byteArrayOf(type.value))

    constructor(type: BleCommandType, payload: ByteArray) : this(
        byteArrayOf(type.value) + payload
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is BleCommand) return false

        if (!data.contentEquals(other.data)) return false

        return true
    }

    override fun toString(): String {
        return "Raw command: [${data.toHex()}]";
    }

    override fun hashCode(): Int {
        return data.contentHashCode()
    }
}

class BleCommandRTS : BleCommand(BleCommandType.RTS)

class BleCommandCTS : BleCommand(BleCommandType.CTS)

class BleCommandAbort : BleCommand(BleCommandType.ABORT)

class BleCommandSuccess : BleCommand(BleCommandType.SUCCESS)

class BleCommandFail : BleCommand(BleCommandType.FAIL)
